Update editor sidebar validation notifications

delawski opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #5304
Relates to #5741

This PR updates the look and feel of the AMP sidebar, most notably the validation notifications area.

List of collapsed errors:

<img width="280" alt="Screenshot 2021-03-17 at 16 07 32" src="https://user-images.githubusercontent.com/478735/111492205-c3efeb00-873c-11eb-9389-08a933bf34bf.png">

Expanded error with a long message:

<img width="281" alt="Screenshot 2021-03-17 at 16 07 46" src="https://user-images.githubusercontent.com/478735/111491366-167cd780-873c-11eb-8826-b56e6fabf7f9.png">

Here are the most important changes included in this PR:
- Extract the logic that determines the messaging in the sidebar from the `<Sidebar />` component into the new `amp-validation-status` components (`<AMPValidationStatusNotification />` and `<AMPRevalidateNotification />`).
- Introduce a generic `<SidebarNotification />` component (along with a dedicated container component).
- Make the error title in the panel title more compact when collapsed and more descriptive when expanded.
- Use the `text-overflow: ellipsis` property for error title when collapsed.
- Remove the now redundant title from the error content area (title in full is already part of the panel title).
- Add an icon to the block type component in the error content area.
- Mark only the "kept" errors with a red left border and remove the exclamation mark from the error panel title.
- Mark new errors with a light grey background and bold font.
- Use a white background and a regular `font-weight` for the reviewed errors.
- Clean up the CSS and use alphabetical ordering for properties.
- Extract `amp-error`-related CSS to a separate stylesheet.
- Use classname-based CSS selectors for the errors list and list items.
- Don't escape error message title so that HTML tags like `<code>` are properly rendered.

The new components are meant to be reused in the Pre-Publish Dialogue and in the general Sidebar, in the AMP section.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

delawski committed:
Introduce editor sidebar notification component

delawski committed:
Introduce AMP status icon to be used in sidebar notifications

delawski committed:
Set `isFetchingErrors` flag when editor loads

delawski committed:
Extract validation status messages to notification components

delawski committed:
Update the reviewed errors toggle appearance

delawski committed:
Use uniform CSS naming convention for Icon and add unit tests

delawski committed:
Do not show revalidate notification when fetching validation errors

delawski committed:
Add unit test to cover validation errors fetching state

westonruter commented:
@delawski FYI: There's a conflict now with `develop`.

delawski commented:
> @delawski FYI: There's a conflict now with `develop`.

Thanks for the heads-up. I'll take care of the conflicts when working on the few updates that are still on my todo list for this PR.

jwold commented:
1. @delawski for error handling what is an example of what an error might say? 
2. Assuming you then need a UI displaying that error. Correct?

delawski commented:
> 1. For error handling what is an example of what an error might say?

Here's an example of an error message returned by the backend:

<img width="750" alt="Screenshot 2021-02-26 at 14 58 04" src="https://user-images.githubusercontent.com/478735/109309041-0e1f3400-7843-11eb-8789-80c825dcb8b9.png">

> 2. Assuming you then need a UI displaying that error. Correct?

Correct.

delawski committed:
Merge branch 'develop' into feature/5304-validation-notifications

# Conflicts:
#	assets/src/block-validation/sidebar.js

delawski committed:
Fix build issue related to SVG markup

On the build step, I was getting the following error thrown by the `block-validation/icon.js`:

```
Property value of ObjectProperty expected node to be of a type ["Expression","PatternLike"] but instead got "JSXExpressionContainer"
```

Cleaning up the `amp-logo-icon.svg` markup and passing it through svgomg fixed the issue.

delawski committed:
Add loading state message and reorder notifications in sidebar

delawski committed:
Add API request error handling

delawski commented:
@westonruter @jwold With 543ab48 and c37125e a loading state and an API errors handling have been added to the sidebar notification area:

https://user-images.githubusercontent.com/478735/110360818-835ae800-803f-11eb-8ae0-1319c8cfd9eb.mov

The error message displayed in the notification is whatever the REST endpoint returns. This might be improved over time. However, since such API-level errors are rather rare I wouldn't worry about that too much at this point.

jwold commented:
This is perfect for now. Great job @delawski !

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/5304-validation-notifications

* 'develop' of github.com:ampproject/amp-wp: (77 commits)
  Summarize cleaned markdown file count
  Indicate WP_CLI::error() among earlyTerminatingMethodCalls
  Revert change to WORDPRESS_DB_USER
  Remove sizereport.config.js
  Rename env file
  Move duplicate env vars to file
  Supply WP related env vars to CLI container as well
  Debug wp config
  Remove --quiet flag
  Use alternative method to execute SQL command
  Create db before ensuring connection to WordPress
  Debug mysql command
  Disable pseudo-tty allocation for docker-compose command
  Use latest WP image for local env; create database during env setup
  Update workflow name
  Use alternative workflow ID
  Rename workflow
  Don't run workflow if certain files change
  Cancel any previous queued or in-progress workflow runs on push to a branch
  Reduce specificity of CSS selector
  ...

westonruter commented:
Sorry for the delay in reviewing. 

@delawski For validation errors which are reviewed, could you add a white left-border or add additional padding so that the unreviewed and reviewed have consistent alignment:

![image](https://user-images.githubusercontent.com/134745/111258107-8d4d8f80-85d9-11eb-98c5-c82df2cc52fc.png)

westonruter commented:
@delawski Also, where did the link go to access the Validated URL screen? There should be a details link still.

westonruter commented:
The error titles can be much longer than just “invalid element”, especially when translated. Currently the client-side logic is doing something hacky by truncating the title at the first colon. So instead of `Invalid element: <code>bad</code>` it becomes just `Invalid element`. This doesn't always work, for example:

> ![image](https://user-images.githubusercontent.com/134745/111258901-216c2680-85db-11eb-950a-e1c26417266d.png)

I think we need to:

- [ ] Discontinue the truncation at the colon.
- [ ] Not have “kept” label spelled out. Having just the icon should be fine because they can expand to understand what the icon means.

Here's a more extreme example of how long the titles can be:

> ![image](https://user-images.githubusercontent.com/134745/111259099-845dbd80-85db-11eb-92c2-296afc89af3e.png)

Naturally the `<code>` markup should not be getting escaped in this scenario.

westonruter commented:
@jwold The left red border in the notification doesn't feel quite right when compared with the left red borders appearing with the validation errors, no? Feels inconsistent to me:

![image](https://user-images.githubusercontent.com/134745/111259522-5036cc80-85dc-11eb-9fcd-3486dcf7b7d5.png)

westonruter commented:
How about rather giving the notice the same red outline that is used with each block that has an unreviewed validation error:

![image](https://user-images.githubusercontent.com/134745/111259683-a73ca180-85dc-11eb-80e1-570505fe4bcf.png)

delawski commented:
> where did the link go to access the Validated URL screen? There should be a details link still.

@westonruter Good catch! It slipped through the cracks, we should get it back in.

@jwold Can you suggest the best location for the "View technical details" link? I think it should land in the more-less previous position, below the status message:

<img width="280" alt="Screenshot 2021-03-16 at 12 03 24" src="https://user-images.githubusercontent.com/478735/111299025-bb6db680-864f-11eb-8b6b-0309a5bbae6f.png">

What do you think?

---

**Update**

I went ahead and added the review link to the status message in e3ebdca:

<img width="280" alt="Screenshot 2021-03-16 at 13 46 19" src="https://user-images.githubusercontent.com/478735/111311418-82890e00-865e-11eb-86ea-db6334eb038a.png">

We can always revisit this in a follow-up commit/PR.

delawski committed:
Add review link back to validation status notification

delawski committed:
Use consistent border-left treatment for error messages

delawski committed:
Use refreshed design for AMP sidebar

delawski commented:
The look and feel of the AMP sidebar have been updated in 74868b9 according to the latest designs.

The key changes are:

- Only the "kept" errors are marked with a red left border. Also, the exclamation mark has been removed from the panel title.
- The new errors are marked with a light grey background and bold font.
- The reviewed errors use a white background and the regular font-weight.
- Errors that refer to removed blocks are greyed out (if an error was marked as "kept", it has the left border but greyed out as well).

Here is a set of various types of errors:

<img width="281" alt="Screenshot 2021-03-16 at 15 33 41" src="https://user-images.githubusercontent.com/478735/111330433-36df6000-8670-11eb-9e58-8dfc686858d3.png">

delawski commented:
For the issue with an error message that's too long, I've played around quickly with `text-overflow: ellipsis` and other CSS tweaks. Here's what I came up with:

<img width="281" alt="Screenshot 2021-03-16 at 16 11 32" src="https://user-images.githubusercontent.com/478735/111332883-75761a00-8672-11eb-9993-19c513dea97e.png">

The message in full would be displayed only after opening an error panel. I've also moved the error type symbol below the icon in order to gain a bit more horizontal space. What do you think, @westonruter @jwold?

jwold commented:
@delawski those are great changes. 

Sidenote: I'm wondering if we need to always show both of these: https://d.pr/i/9tuFat. Early on we wanted to help provide additional information as folks scanned through the error list, but now I'm just re-evaluating what benefit it provides as folks scroll through.

delawski commented:
> Sidenote: I'm wondering if we need to always show both of these: https://d.pr/i/9tuFat. Early on we wanted to help provide additional information as folks scanned through the error list, but now I'm just re-evaluating what benefit it provides as folks scroll through.

I agree it may seem redundant and not very informative. I guess we could drop that element. What do others think?

westonruter commented:
> * Only the "kept" errors are marked with a red left border. Also, the exclamation mark has been removed from the panel title.

Good. That is much more concise.

> * The new errors are marked with a light grey background and bold font.
> * The reviewed errors use a white background and the regular font-weight.

Good, as this aligns with the email design pattern.

The only issue with the previous changes is they now are inconsistent with how errors are displayed on the Validated URL, Validated URLs, and Error Index screens:

![image](https://user-images.githubusercontent.com/134745/111355369-8c057c80-8644-11eb-841f-d52ae12674e5.png)

![image](https://user-images.githubusercontent.com/134745/111355409-96c01180-8644-11eb-9934-8cdfae83f9a1.png)

![image](https://user-images.githubusercontent.com/134745/111355455-9f184c80-8644-11eb-8657-dd19b0dfc322.png)

Should these be updated to use the same styling? This could be done in a separate issue/PR.

> For the issue with an error message that's too long, I've played around quickly with `text-overflow: ellipsis` and other CSS tweaks. Here's what I came up with:

I like that. However just make sure that the `<code>` gets replaced with actual `code` elements. Showing the full error upon expansion seems like a good idea.

> I agree it may seem redundant and not very informative. I guess we could drop that element. What do others think?

If the label “HTML” is removed the then the “error type” could be displayed when the error is expanded. As long as the HTML, JS, and CSS error icons are clear enough, then I think the label can be dropped.

delawski committed:
Revamp sidebar AMP errors list CSS

- Extract `amp-error` related CSS to a separate stylesheet.
- Clean up the CSS and use alphabetical ordering for properties.
- Make the error title in the panel title more compact when collapsed and more descriptive when expanded.
- Use classname-based CSS selectors for the errors list and list items.
- Don't escape error message title so that HTML tags like `<code>` are properly rendered.
- Use the `text-overflow: ellipsis` property for error title when collapsed.
- Remove the now redundant title from the error content area (title in full is already part of the panel title).
- Add icon to the block type component in the error content area.
- Fix unit tests.

delawski commented:
In e442893 I've addressed the latest feedback and also did some cleanup (mostly CSS). Here's a list of changes:

- Extract `amp-error` related CSS to a separate stylesheet.
- Clean up the CSS and use alphabetical ordering for properties.
- Make the error title in the panel title more compact when collapsed and more descriptive when expanded.
- Use classname-based CSS selectors for the errors list and list items.
- Don't escape error message title so that HTML tags like `<code>` are properly rendered.
- Use the `text-overflow: ellipsis` property for error title when collapsed.
- Remove the now redundant title from the error content area (title in full is already part of the panel title).
- Add an icon to the block type component in the error content area.
- Fix unit tests.

Here's a preview of collapsed errors list:

<img width="280" alt="Screenshot 2021-03-17 at 16 07 32" src="https://user-images.githubusercontent.com/478735/111490465-57282100-873b-11eb-8036-8931cf485456.png">

And here's a screenshot of an expanded error that has a long title:

<img width="281" alt="Screenshot 2021-03-17 at 16 07 46" src="https://user-images.githubusercontent.com/478735/111490571-6dce7800-873b-11eb-8521-08091fd9f55f.png">

> The only issue with the previous changes is they now are inconsistent with how errors are displayed on the Validated URL, Validated URLs, and Error Index screens

@westonruter Yes, that's something we should do next. And I agree that we should do it in a separate PR and not "pollute" this (already hefty) PR.

---

@jwold What do you think about @westonruter point in https://github.com/ampproject/amp-wp/pull/5929#issuecomment-799957644?

westonruter commented:
> @jwold What do you think about @westonruter point in [#5929 (comment)](https://github.com/ampproject/amp-wp/pull/5929#issuecomment-799957644)?

Then again, this border is redundant since the icon already communicates the state, no? In that case the border can be removed entirely for the notification.

jwold commented:
> Then again, this border is redundant since the icon already communicates the state, no? In that case the border can be removed entirely for the notification.

That works for me as well.

westonruter commented on assets/src/block-editor/components/sidebar-notification/style.css:
```diff
@@ -0,0 +1,70 @@
+	position: relative;
+}
+
+.sidebar-notification.is-error::before {
+	background: #d64e21;
+	bottom: 0;
+	content: "";
+	left: 0;
+	position: absolute;
+	top: 0;
+	width: 5px;
+}
+
```
```suggestion
```

![image](https://user-images.githubusercontent.com/134745/111580976-b1de6e80-8775-11eb-86c9-4cc03864ad6b.png)

delawski replied:
The entire `isError` component state can be removed since it's the only instance where it's used.

It's been done in 6b78c8b.

westonruter commented on assets/src/block-editor/components/sidebar-notification/index.js:
```diff
@@ -0,0 +1,82 @@
+ * Notification component used in the block editor sidebar.
+ *
+ * @param {Object} props
+ * @param {string|Object} props.action Call to action element.
+ * @param {string|Object} props.icon Status icon element.
```
Are strings ever really passed here?

```suggestion
 * @param {Object} props.action Call to action element.
 * @param {Object} props.icon Status icon element.
```

westonruter committed:
Clarify what outside content means

westonruter commented on assets/src/block-validation/sidebar.js:
```diff
@@ -74,113 +49,36 @@ export function Sidebar() {
-				</>
+			{ 0 < displayedErrors.length && (
+				<ul className="amp-sidebar__errors-list">
+					{ displayedErrors.map( ( validationError ) => (
+						<li
+							key={ validationError.clientId }
```
When I have a validation error coming from outside the content, I'm getting a React error:

![image](https://user-images.githubusercontent.com/134745/111583513-cf153c00-8779-11eb-8180-934e8232ba27.png)

Something like this seems necessary for errors that aren't connected with blocks:

```suggestion
					{ displayedErrors.map( ( validationError, i ) => (
						<li
							key={ validationError.clientId || i }
```

delawski replied:
I've addressed this issue in 9d35150. Initially, I didn't take into consideration the fact that some errors may have no `clientId` property (since they don't come from blocks).

westonruter reviewed (changes requested):
I'm seeing a strange issue where when I switch to the code editor view, all blocks say (incorrectly) that they have been removed from the document:

https://user-images.githubusercontent.com/134745/111583606-f53adc00-8779-11eb-936c-131876d92df4.mov

If I save draft, then they all get restored in the sidebar.

delawski committed:
Add index back to errors list `key` in case there is no `clientId`

delawski committed:
Remove redundant error state from status notification

delawski committed:
Update assets/src/block-editor/components/sidebar-notification/index.js

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Fix broken unit tests and update snapshots

delawski committed:
Merge remote-tracking branch 'origin/feature/5304-validation-notifications' into feature/5304-validation-notifications

delawski commented:
It seems the Code Editor bug is something we have on `develop`, too:

![Screenshot 2021-03-18 at 11 27 39](https://user-images.githubusercontent.com/478735/111611770-1aad0180-87dd-11eb-8e40-783785e80dc8.png)

I'll debug it today.

westonruter commented:
@delawski Regarding the issue of the incorrectly-reported removed blocks, how about changing the messaging for now to not say the block was removed but rather that the error can't currently be found and that the post should be revalidated? This would address the immediate confusion while investigating the separate issue with the changing client IDs.

delawski commented:
> Regarding the issue of the incorrectly-reported removed blocks, how about changing the messaging for now to not say the block was removed but rather that the error can't currently be found and that the post should be revalidated? This would address the immediate confusion while investigating the separate issue with the changing client IDs.

@westonruter That seems like a reasonable approach. I can push the update tomorrow.

I'd also want to check tomorrow if there is maybe an easy fix for the (now separate new issue) #5994.

pierlon commented on assets/src/block-editor/components/sidebar-notification/index.js:
```diff
@@ -0,0 +1,74 @@
+ * Notification component used in the block editor sidebar.
+ *
+ * @param {Object} props
+ * @param {Object} props.action Call to action element.
+ * @param {Object} props.icon Status icon element.
```
The `ReactElement` type would suit these parameters better.

(Note that `ReactElement` would have to be imported from `react`, as suggested below).

delawski replied:
Good catch.

It actually seems that using a `ReactChild` might be even better since both we accept either elements or strings here:

```js
type ReactText = string | number;
type ReactChild = ReactElement | ReactText;
```

What do you think, @pierlon?

delawski replied:
@pierlon I've addressed this one in efeba44. I've used the `ReactChild` but feel free to amend if you think another type is better-suited.

pierlon commented on assets/src/block-editor/components/sidebar-notification/index.js:
```diff
@@ -0,0 +1,74 @@
+ * External dependencies
+ */
+import classnames from 'classnames';
+import PropTypes from 'prop-types';
```
Following comment from above.

```suggestion
import PropTypes from 'prop-types';
import { ReactElement } from 'react';
```

delawski replied:
Addressed in efeba44.

westonruter committed:
Show revalidate notification when editor mode has switched

westonruter commented:
> That seems like a reasonable approach. I can push the update tomorrow.

@delawski I took a stab at this. 

Before | After
-------|-------
<img width="281" alt="Screen Shot 2021-03-18 at 12 38 36" src="https://user-images.githubusercontent.com/134745/111692169-61eebe80-87ec-11eb-85ed-ac5622f30e8d.png"> | <img width="282" alt="Screen Shot 2021-03-18 at 13 15 31" src="https://user-images.githubusercontent.com/134745/111692178-631feb80-87ec-11eb-940c-7ed84681e070.png">

You can see that it shows the “maybe dirty” notification when merely validation errors occur for blocks that seem to be removed, but then when the editor truly _does_ become dirty, it then says so:

https://user-images.githubusercontent.com/134745/111692362-982c3e00-87ec-11eb-80d2-f2b7e5d2e39e.mov

Please amend 74c6e37 as you see fit.

pierlon commented on assets/src/block-editor/components/sidebar-notification/index.js:
```diff
@@ -0,0 +1,74 @@
+ * @param {string} props.message Message text.
+ */
+export function SidebarNotification( {
+	action,
+	icon,
```
As stated in the [JSX docs](https://reactjs.org/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized), user-defined components should be capitalized.


```suggestion
	Action,
	Icon,
```

delawski replied:
I think this applies only to the component definition, not necessarily to the case when a component is passed via props. If such a prop would be capitalized, then it would also have to be capitalized when using a component and it doesn't seem right, i.e.

```js
<SidebarNotification Action={ ... } Icon={ ... } />
```

The official [docs](https://reactjs.org/docs/composition-vs-inheritance.html#containment) don't suggest using capitalized props in the `SplitPane` example:

```js
function SplitPane(props) {
  return (
    <div className="SplitPane">
      <div className="SplitPane-left">
        {props.left}
      </div>
      <div className="SplitPane-right">
        {props.right}
      </div>
    </div>
  );
}

function App() {
  return (
    <SplitPane
      left={
        <Contacts />
      }
      right={
        <Chat />
      } />
  );
}
```

In Gutenberg, such props are starting with a lowercase letter, too (check out an [example](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/placeholder/index.js#L25-L26)).

pierlon replied:
Ah right that makes sense. Thanks for the examples.

westonruter committed:
Fix test after message change in 74c6e3798

pierlon commented on assets/src/block-editor/components/sidebar-notification/index.js:
```diff
@@ -0,0 +1,74 @@
+		<div className={ classnames( 'sidebar-notification', { 'is-loading': isLoading } ) }>
+			{ iconElement && (
+				<div className="sidebar-notification__icon">
+					{ iconElement }
```
Any reason this isn't using the JSX syntax (i.e `<iconElement />`?

delawski replied:
For conditionally rendered components, the [React docs suggest using element variables](https://reactjs.org/docs/conditional-rendering.html#element-variables). The same convention is [used](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/gallery/save.js#L40-L51) in Gutenberg as well.

If we wanted to go with a JSX syntax, I guess we should [capitalize the component name](https://reactjs.org/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized) but then the condition would _look_ a bit off:

```diff
- { iconElement && (
+ { IconElement && (
```

pierlon replied:
Yea let's keep what you have.

pierlon commented on assets/src/block-editor/components/sidebar-notification/index.js:
```diff
@@ -0,0 +1,74 @@
+				</p>
+				{ action && (
+					<p>
+						{ action }
```
Same question as above.

pierlon commented on assets/src/block-validation/amp-validation-status/revalidate-notification.js:
```diff
@@ -0,0 +1,105 @@
+		isPostDirty,
+	} = useSelect( ( select ) => ( {
+		hasActiveMetaboxes: select( 'core/edit-post' ).hasMetaBoxes(),
+		isDraft: [ 'draft', 'auto-draft' ].indexOf( select( 'core/editor' )?.getEditedPostAttribute( 'status' ) ) !== -1,
```
Any reason for using the optional chaining operator here? Would `getEditedPostAttribute()` not always be available?

delawski replied:
Addressed in cb830a0.

pierlon commented on assets/src/block-validation/amp-validation-status/status-notification.js:
```diff
@@ -0,0 +1,107 @@
+		ampCompatibilityBroken: select( BLOCK_VALIDATION_STORE_KEY ).getAMPCompatibilityBroken(),
+		fetchingErrorsRequestErrorMessage: select( BLOCK_VALIDATION_STORE_KEY ).getFetchingErrorsRequestErrorMessage(),
+		hasValidationErrors: select( BLOCK_VALIDATION_STORE_KEY ).getValidationErrors()?.length > 0,
+		isDraft: [ 'draft', 'auto-draft' ].indexOf( select( 'core/editor' )?.getEditedPostAttribute( 'status' ) ) !== -1,
```
Any reason for using the optional chaining operator here? Would `getEditedPostAttribute()` not always be available?

delawski replied:
You're right – we don't need the optional chaining here.

delawski replied:
Addressed in cb830a0.

pierlon commented on assets/src/block-validation/amp-validation-status/status-notification.js:
```diff
@@ -0,0 +1,107 @@
+	} = useSelect( ( select ) => ( {
+		ampCompatibilityBroken: select( BLOCK_VALIDATION_STORE_KEY ).getAMPCompatibilityBroken(),
+		fetchingErrorsRequestErrorMessage: select( BLOCK_VALIDATION_STORE_KEY ).getFetchingErrorsRequestErrorMessage(),
+		hasValidationErrors: select( BLOCK_VALIDATION_STORE_KEY ).getValidationErrors()?.length > 0,
```
The `getValidationErrors` selector would always return an array, no?

westonruter replied:
I removed the `?` as part of 8b42ad96e385431f71f561818126ef8d0fcd8cd9

pierlon commented on assets/src/block-validation/amp-validation-status/status-notification.js:
```diff
@@ -0,0 +1,107 @@
+		return (
+			<SidebarNotification
+				icon={ <AMPValidationErrorsKeptIcon /> }
+				message={ __( 'AMP blocked from validation issues marked kept.', 'amp' ) }
```
The wording may not be clear to the user here. How about something like:

```suggestion
				message={ __( 'AMP is disabled due to validation issues being marked as kept.', 'amp' ) }
```

cc @westonruter

westonruter replied:
How about rather:

```suggestion
				message={ __( 'AMP is disabled due to invalid markup being kept.', 'amp' ) }
```

westonruter replied:
The issue count should probably be indicated as well.

westonruter replied:
See 8b42ad96e385431f71f561818126ef8d0fcd8cd9

pierlon commented on assets/src/block-validation/amp-validation-status/status-notification.js:
```diff
@@ -0,0 +1,107 @@
+		return (
+			<SidebarNotification
+				icon={ <StatusIcon broken={ true } /> }
+				message={ __( 'AMP is working, but issues needs review.', 'amp' ) }
```
Similar to above.

```suggestion
				message={ __( 'AMP is enabled, but there are issues that need to be reviewed.', 'amp' ) }
```

cc @westonruter

westonruter replied:
Slight iteration on that:

```suggestion
				message={ __( 'AMP is enabled, but there are issues needing review.', 'amp' ) }
```

westonruter replied:
Also, we should indicate the number of issues, since if there is only one issue this will be confusing.

westonruter replied:
See 8b42ad96e385431f71f561818126ef8d0fcd8cd9

pierlon commented on assets/src/block-validation/amp-validation-status/status-notification.js:
```diff
@@ -0,0 +1,107 @@
+	return (
+		<SidebarNotification
+			icon={ <StatusIcon /> }
+			message={ __( 'AMP is working. All issues reviewed or removed.', 'amp' ) }
```
Similar to above:

```suggestion
			message={ __( 'AMP is enabled. All issues have been reviewed.', 'amp' ) }
```

cc @westonruter

westonruter replied:
Actually, I think this should be split into two. If there are no validation errors at all, then the message should be: “AMP is enabled. There are no validation issues.”

Otherwise, if there are validation issues but they have all been reviewed, it should be: “AMP is enabled. All validation issues have been reviewed.”

westonruter replied:
Done in 8b42ad96e385431f71f561818126ef8d0fcd8cd9

pierlon reviewed (changes requested):
Almost done reviewing, but submitting what I've found so far.

westonruter committed:
Add error counts to notification messages; make clear when no issues

westonruter committed:
Remove reference to 'post' in pending-save message

westonruter commented on assets/src/block-validation/amp-validation-status/status-notification.js:
```diff
@@ -0,0 +1,154 @@
+		);
+	}
+
+	if ( isEditedPostNew ) {
+		return (
+			<SidebarNotification
+				icon={ <StatusIcon /> }
+				message={ __( 'Validation will be checked upon saving.', 'amp' ) }
+			/>
+		);
+	}
```
Would it make sense to move this condition up to be the first after the `isFetchingErrors` conditional?

pierlon replied:
That makes sense, yes.

delawski replied:
Done in 56ceb8c.

pierlon commented:
Would it make sense adding titles for each section? At the moment they seem bare and the user may not immediately know what they're looking at.

![image](https://user-images.githubusercontent.com/16200219/111723977-5509b380-885c-11eb-9a02-e0614d13a0ca.png)

pierlon commented:
While testing this I found given a HTML block with the content `<?php`, a validation error is only raised when it is the first block on the page, which is weird. Here's an example:


https://user-images.githubusercontent.com/16200219/111726116-32799980-8860-11eb-9a00-8553b208966b.mp4

pierlon commented:
Bug replicated on the `develop` branch as well, so it's not from this PR. Will investigate.

pierlon commented on assets/src/block-validation/sidebar.js:
```diff
@@ -74,113 +49,37 @@ export function Sidebar() {
-					</div>
-					{ hasReviewedValidationErrors && (
-						<div className="amp-sidebar__options">
-							<ToggleControl
-								checked={ isShowingReviewed }
-								label={ __( 'Include reviewed issues', 'amp' ) }
-								onChange={ ( newIsShowingReviewed ) => {
-									setIsShowingReviewed( newIsShowingReviewed );
-								} }
-							/>
-						</div>
```
Why was this toggle removed? The current replacement seems like a downgrade.

Before | After
---|---
![image](https://user-images.githubusercontent.com/16200219/111727168-47572c80-8862-11eb-9177-f0ab25231393.png) | ![image](https://user-images.githubusercontent.com/16200219/111727553-feec3e80-8862-11eb-9775-d1e2cf068d55.png)

westonruter replied:
Yes, because the toggle seems to prominent. The link feels a bit more discrete.

pierlon replied:
Ah OK.

pierlon commented on assets/src/block-validation/sidebar.js:
```diff
@@ -74,113 +49,37 @@ export function Sidebar() {
-						</PanelBody>
-					)
+			{ hasReviewedValidationErrors && (
+				<div className="amp-sidebar__options">
```
Would it be better to place this above the list of errors? One would potentially have to scroll all the way down to see this button (due to viewport constraints or having many errors).

westonruter replied:
The idea is that this would be clicked infrequently. The main thing a user will care about are the unreviewed issues. The link to show all is there just for curiosity.

westonruter commented:
> Would it make sense adding titles for each section? At the moment they seem bare and the user may not immediately know what they're looking at.

Maybe the entire sidebar should be called "AMP Validation" instead of just "AMP" for now?

pierlon commented:
> Maybe the entire sidebar should be called "AMP Validation" instead of just "AMP" for now?

That could work as well.

westonruter commented:
> While testing this I found given a HTML block with the content `<?php`, a validation error is only raised when it is the first block on the page, which is weird. Here's an example:

@pierlon It's simply because your `<script>` tag was not closed properly. If I look at the non-AMP version the syntax highlighting indicates as such: 

![image](https://user-images.githubusercontent.com/134745/111737432-d9e4d500-883c-11eb-9632-6bebde1e2db1.png)

Also see this example code:

```php
<?php

use AmpProject\Dom\Document;

$script_first = '<div>
	<script</script>
	<?php
	?>
</div>';

$pi_first = '<div>
	<?php
	?>
	<script</script>
</div>';

$script_first_doc = Document::fromHtmlFragment( $script_first );
$pi_first_doc = Document::fromHtmlFragment( $pi_first );

echo "## script_first_doc\n";
echo $script_first_doc->saveHTML( $script_first_doc->body );

echo "\n\n";

echo "## pi_first_doc\n";
echo $pi_first_doc->saveHTML( $pi_first_doc->body );
```

Output:

```html
## script_first_doc
<body><div>
	<script>
	<?php
	?>
</script></div></body>

## pi_first_doc
<body><div>
	<?php ?>
	<script>
</script></div></body>
```

Notice the PHP processing instruction gets rendered _inside_ the `script` in the first example, and the PHP PI is there considered text content not markup. So it is not a separate validation error.

westonruter committed:
Change block editor plugin title from 'AMP' to 'AMP Validation'

westonruter commented:
> > Maybe the entire sidebar should be called "AMP Validation" instead of just "AMP" for now?

I've made this change:

<img width="325" alt="Screen Shot 2021-03-18 at 23 01 45" src="https://user-images.githubusercontent.com/134745/111738232-35639280-883e-11eb-9e12-dd8bee8980c1.png"> <img width="328" alt="Screen Shot 2021-03-18 at 23 02 01" src="https://user-images.githubusercontent.com/134745/111738236-372d5600-883e-11eb-886d-5220ebebe35f.png">

In the future when/if we add non-validation stuff to the sidebar, we can change it back to "AMP".

westonruter commented:
@delawski Would you open a new issue for adding the pre-publish panel?

delawski committed:
Remove unnecessary optional chaining operator

delawski committed:
Use more exact types in JSDoc and propTypes definition

delawski committed:
Get `blockName` directly without other properties

delawski committed:
Check if post is new early

delawski commented:
@pierlon @westonruter I've addressed your feedback. I think the only thing left to do is to update the unit tests for the `StatusNotification` which I'll do shortly.

> Would you open a new issue for adding the pre-publish panel?

Yes, I'll create a single ticket for the pre-publish panel and for the panel in the document sidebar.

delawski commented on assets/src/block-validation/amp-validation-status/revalidate-notification.js:
```diff
@@ -0,0 +1,121 @@
+		let _hasErrorsFromRemovedBlocks = false;
+		for ( const validationError of select( BLOCK_VALIDATION_STORE_KEY ).getValidationErrors() ) {
+			const { clientId } = validationError;
+			const blockDetails = clientId ? select( 'core/block-editor' ).getBlock( clientId ) : null;
+			if ( clientId && ! blockDetails ) {
```
As in aa668d0, we could get here just the block name instead of the entire block object, i.e.

```suggestion
			if ( clientId && ! select( 'core/block-editor' ).getBlockName( clientId ) ) {
```

delawski replied:
When I think about it, we could actually replace the entire loop with `find` method like

```diff
-		let _hasErrorsFromRemovedBlocks = false;
-		for ( const validationError of select( BLOCK_VALIDATION_STORE_KEY ).getValidationErrors() ) {
-			const { clientId } = validationError;
-			const blockDetails = clientId ? select( 'core/block-editor' ).getBlock( clientId ) : null;
-			if ( clientId && ! blockDetails ) {
-				_hasErrorsFromRemovedBlocks = true;
-				break;
-			}
-		}
-
-		return {
-			hasErrorsFromRemovedBlocks: _hasErrorsFromRemovedBlocks,
+		const errors = select( BLOCK_VALIDATION_STORE_KEY ).getValidationErrors();
+
+		return {
+			hasErrorsFromRemovedBlocks: Boolean( errors.find( ( { clientId } ) => clientId && ! select( 'core/block-editor' ).getBlockName( clientId ) ) ),
```

But I guess it might not be as readable as the loop.

westonruter replied:
I like that latter suggestion. Go ahead with it!

delawski replied:
Sounds good, it's done (in an even shorter form) in 17ef326.

delawski committed:
Fix unit tests and update snapshots

delawski commented:
The unit tests should now be good (see: 4794812).

> > Would you open a new issue for adding the pre-publish panel?
>
> Yes, I'll create a single ticket for the pre-publish panel and for the panel in the document sidebar.

Here's the follow-up ticket: #5997

delawski committed:
Use array find method instead of loop

westonruter reviewed (approved):
A thing of beauty!

_It is a gift!_

pierlon commented:
> @pierlon It's simply because your <script> tag was not closed properly...

Ah that makes sense. Thanks for solving the mystery @westonruter :mag:.

pierlon commented on assets/src/block-validation/amp-validation-status/revalidate-notification.js:
```diff
@@ -0,0 +1,111 @@
+		isEditedPostNew,
+		isPostDirty,
+	} = useSelect( ( select ) => ( {
+		hasErrorsFromRemovedBlocks: Boolean( select( BLOCK_VALIDATION_STORE_KEY ).getValidationErrors().find(
+			( { clientId } ) => clientId && ! select( 'core/block-editor' ).getBlockName( clientId ) ),
+		),
```
Nicely done :+1:

pierlon reviewed (approved):
Marvelous work done here :trophy:.
